gitlab-status/handler.go

```diff
@@ -0,0 +1,132 @@
+	"os"
+	"strings"
+
+	"github.com/alexellis/hmac"
+	"github.com/openfaas/openfaas-cloud/sdk"
+)
+
+// Handle a serverless request
+func Handle(req []byte) string {
+	validateError := validateRequest(req)
+	if validateError != nil {
+		log.Fatal(validateError)
+	}
+
+	status, statusErr := sdk.UnmarshalStatus(req)
+	if statusErr != nil {
+		return statusErr.Error()
+	}
+
+	token, tokenErr := sdk.ReadSecret("gitlab-api-token")
+	if tokenErr != nil {
+		return tokenErr.Error()
+	}
+
+	url := gitLabURLBuilder(status)
+
+	for _, commitStatus := range status.CommitStatuses {
+		reportErr := sendReport(url, token, commitStatus.Status, commitStatus.Description, commitStatus.Context)
+		if reportErr != nil {
+			log.Fatalf("failed to report status %v, error: %s", status, reportErr.Error())
+		}
+	}
+
+	return ""
+}
+
+func gitLabURLBuilder(status *sdk.Status) string {
+	delimeterPlace := 3
+	delimeterValue := "/"
+	baseURL := getURLbyDelimeter(status.EventInfo.URL, delimeterPlace, delimeterValue)
+	routeURL := fmt.Sprintf("/api/v4/projects/%d/statuses/%s?", status.EventInfo.InstallationID, status.EventInfo.SHA)
+	newURL := append(strings.Split(baseURL, ""), routeURL)
+	wholeURL := strings.Join(newURL, "")
+	return wholeURL
+}
+
+func getURLbyDelimeter(baseURL string, position int, delimeter string) string {
+	sliceURL := strings.Split(baseURL, "")
+	var slash, last int
+	for index, symbol := range sliceURL {
+		if slash == position {
+			last = index
+			break
+		}
+		if symbol == delimeter {
+			slash++
+		}
+	}
+
+	sliceURL = append(sliceURL[:last-1], sliceURL[cap(sliceURL):]...)
+	baseURL = strings.Join(sliceURL, "")
+	return baseURL
+}
+
+func appendParameters(URL string, state string, desc string, context string) (string, error) {
+	var theURL *url.URL
+
+	theURL, urlErr := url.Parse(URL)
+	if urlErr != nil {
+		return "", urlErr
+	}
+
+	if state == "failure" {
+		state = "failed"
+	}
+
+	parameters := url.Values{}
+	parameters.Add("state", state)
+	parameters.Add("description", desc)
+	parameters.Add("context", context)
+	theURL.RawQuery = parameters.Encode()
+
+	return theURL.String(), nil
+
+}
+
+func sendReport(URL string, token string, state string, desc string, context string) error {
+	fullURL, fullURLErr := appendParameters(URL, state, desc, context)
+	if fullURLErr != nil {
+		return fullURLErr
+	}
+	var b io.Reader
+	req, reqErr := http.NewRequest("POST", fullURL, b)
+	if reqErr != nil {
+		return reqErr
+	}
+	req.Header.Set("PRIVATE-TOKEN", token)
+	client := &http.Client{}
+	resp, clientErr := client.Do(req)
+	if clientErr != nil {
+		return clientErr
+	}
+	resp.Body.Close()
+
+	return nil
+}
+
+func validateRequest(req []byte) (err error) {
+	payloadSecret, err := sdk.ReadSecret("payload-secret")
+
+	if err != nil {
+		return fmt.Errorf("couldn't get payload-secret: %t", err)
+	}
+
+	xCloudSignature := os.Getenv("Http_X_Cloud_Signature")
+
+	err = hmac.Validate(req, xCloudSignature, payloadSecret)
+
+	if err != nil {
+		return err
+	}
```

Comment:
It should be added to sdk, but @martindekov is saying that this function appears everywhere. I would write it also:
```
if err := hmac.Validate(req, xCloudSignature, payloadSecret); err != nil {
 	return err
}
```

Reply:
Can be addressed in another PR

Reply:
Amended fix for this